SmartDeviceLink/SDLIAPSession.m

```diff
@@ -64,11 +86,126 @@ - (BOOL)start {
+}
+
+- (BOOL)sdl_dequeueAndWriteToOutputStream:(NSError **)error{
+    NSOutputStream *ostream = self.easession.outputStream;
+    NSMutableData *remainder = [self.sendDataQueue front];
+    BOOL allDataWritten = NO;
+    
+    if (error != nil && remainder != nil && ostream.streamStatus == NSStreamStatusOpen) {
+        NSInteger bytesRemaining = remainder.length;
+        NSInteger bytesWritten = [ostream write:remainder.bytes maxLength:bytesRemaining];
+        if (bytesWritten < 0) {
+            *error = ostream.streamError;
+        } else {
+            if (bytesWritten == bytesRemaining) {
+                // Remove the data from the queue
+                [self.sendDataQueue pop];
+                allDataWritten = YES;
+            } else{
+                // Cleave the sent bytes from the data, the remainder will sit at the head of the queue
+                [remainder replaceBytesInRange:NSMakeRange(0, bytesWritten) withBytes:NULL length:0];
+            }
+        }
+    }
+    
+    return allDataWritten;
+}
+
+- (void)sdl_handleOutputStreamWriteError:(NSError *)error{
+    NSString *errString = [NSString stringWithFormat:@"Output stream error: %@", error];
+    [SDLDebugTool logInfo:errString];
+    // REVIEW: We should look at the domain and the code as a tuple and decide
+    // how to handle the error based on both values. For now, if the stream
+    // is closed, we will flush the send queue and leave it as-is otherwise
+    // so that temporary error conditions can be dealt with by retrying
+    if (self.easession == nil ||
+        self.easession.outputStream == nil ||
+        self.easession.outputStream.streamStatus != NSStreamStatusOpen) {
+        [self.sendDataQueue flush];
+    }
+}
+
+- (void)sdl_accessoryEventLoop {
+    
+  @autoreleasepool {
+    NSAssert(self.easession, @"_session must be assigned before calling");
+    
+    if (!self.easession) {
+      return;
+    }
+    
+    // Open I/O streams of the iAP session
+    NSInputStream *inStream = [self.easession inputStream];
+    [inStream setDelegate:self.streamDelegate];
+    [inStream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
+    [inStream open];
+    
+    NSOutputStream *outStream = [self.easession outputStream];
+    [outStream setDelegate:self.streamDelegate];
+    [outStream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
+    [outStream open];
+    
+    [SDLDebugTool logInfo:@"starting the event loop for accessory"];
+    do {
+        if (self.sendDataQueue.count > 0 && !self.sendDataQueue.frontDequeued) {
+            NSError *sendErr = nil;
+            if (![self sdl_dequeueAndWriteToOutputStream:&sendErr] && sendErr != nil) {
+                [self sdl_handleOutputStreamWriteError:sendErr];
+            }
+        }
+        
+        [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode
+                                 beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.25f]];
```

Comment:
The reasoning is only to block in runMode for a maximum of 0.25s waiting for a stream event. It is not a delay, it is a timeout similar to a try wait or a dispatch wait with a timeout.